### DISCUSSION TITLE
Load only subset of UserAgent data?

### DISCUSSION THREAD

**Discussion by radarsh** [nielsbasjes/yauaa#1554] [Q&A] [answered]
For my case, I know for sure that the requests to my application will only come from desktop browsers and only a handful of browser types as these are the only ones allowed by our enterprise organisation. Is there anyway to instruct YAUAA to only load specific YAML files corresponding to these classes of devices in order to be more memory efficient? For example, might I somehow tell YAUAA to only be prepared to parse Google Chrome user agent strings from Linux, MacOS and Windows?

**Comment by nielsbasjes** [comment, marked as answer]
No that feature is not there out of the box.

There are so many edge cases that pulling the rulesets apart into what you describe would be really hard. The tricky thing is even for just Chrome the patterns have changed over time. So depending on the version things will be structured differently and at some point you can no longer conclude the "Windows 10" marker actually means it was running Windows 10.

You can easily tell Yauaa you are only interested in specified fields, that will automatically drop rules that are not needed to achieve these fields (i.e. some speedup and memory footprint reduction).

If you really want to go bare minimum you can start Yauaa without loading the default ruleset and feed it you own rules.

Building these rules is very hard.
One way would be something like:
- Copy all rule/test files to your own project
- Remove all the test cases you do not care about
- Possibly remove all the fields in the testcases you do not care about
- Then step by step start removing fields and matchers until it is what you want ... (and this will be very very hard to get right)

**Reply by radarsh** [reply]
Thank you. I stumbled upon the `dropDefaultResources()` and `addOptionalResources()` methods against the builder but soon realised that it won't work unless I copied over many default resources into a folder outside the JAR file, which means extra overhead for the project. 

Since I only anticipate a handful of browser types from a handful of operating systems, I'm thinking of doing something like this:

```java
    private final UserAgentAnalyzer userAgentAnalyzer =
        UserAgentAnalyzer
            .newBuilder()
            .withCache(50)
            .withFields(UserAgent.DEVICE_CLASS, UserAgent.AGENT_NAME_VERSION, UserAgent.OPERATING_SYSTEM_NAME_VERSION)
            .build();
```

I observed that if I don't perform a `preheat`, the startup time improves and memory footprint is significantly better. However, I'm not 100% what I would be missing out if I don't perform the `preheat` operation.

**Reply by nielsbasjes** [reply]
The preheat simply runs a lot of the test cases through the engine, giving the jit compiler the chance to do native compilation and preparing more optimized code.
The effect is that once you feed the real data, the latency should be a tad lower.
This step is only useful in real-time applications.

**Reply by radarsh** [reply]
Excellent and that's good to know. Thanks once again and well done for building and maintaining a brilliant library.
